File: server/oidc-boundary.ts
import {
  accountIdDecoder,
  type AccountId,
  type VaultContext,
} from '../lib/domain/identity';
import {
  oidcRedirectUriDecoder,
  oidcStateDecoder,
  oidcNonceDecoder,
  pkceCodeChallengeDecoder,
  pkceCodeVerifierDecoder,
  type OidcAuthorizationCode,
  type OidcClientId,
  type OidcRedirectUri,
  type OidcState,
  type OidcEmailAddress,
  type PkceCodeVerifier,
} from '../lib/domain/oidc';
import {
  createOidcAuthorizationRequest,
  createOidcTransaction,
  decideOidcIdentityResolution,
  oidcCallbackDecoder,
  oidcIdentityRecordDecoder,
  oidcProviderConfigurationDecoder,
  pendingOidcTransactionDecoder,
  validateOidcClaims,
  validateOidcTransaction,
  verifiedOidcClaimsDecoder,
  type OidcAuthorizationRequest,
  type OidcIdentityKey,
  type OidcIdentityRecord,
  type OidcIdentityResolution,
  type PendingOidcTransaction,
} from './core/oidc';

export type OidcClockPort = {
  nowEpochSeconds: () => unknown;
};

export type OidcSecretPort = {
  createState: () => Promise<unknown>;
  createNonce: () => Promise<unknown>;
  createCodeVerifier: () => Promise<unknown>;
};

export type PkceChallengePort = {
  deriveS256: (verifier: PkceCodeVerifier) => Promise<unknown>;
};

export type OidcTransactionStore = {
  insertPending: (transaction: PendingOidcTransaction) => Promise<void>;
  /** Atomically returns and consumes a transaction. A second call must miss. */
  consumeByState: (state: OidcState) => Promise<unknown>;
};

export type OidcCodeExchangeInput = {
  readonly code: OidcAuthorizationCode;
  readonly clientId: OidcClientId;
  readonly redirectUri: OidcRedirectUri;
  readonly codeVerifier: PkceCodeVerifier;
};

export type OidcVerifiedClaimsPort = {
  /**
   * Production adapters must exchange the code over TLS and verify the ID
   * token signature, algorithm, key source, and provider metadata before
   * returning claims. The returned external value is decoded again here.
   */
  exchangeCodeForVerifiedClaims: (
    input: OidcCodeExchangeInput,
  ) => Promise<unknown>;
};

export type OidcIdentityDirectory = {
  findByIssuerSubject: (key: OidcIdentityKey) => Promise<unknown>;
  findAccountIdByVerifiedEmail: (email: OidcEmailAddress) => Promise<unknown>;
};

export type OidcStartIntent = 'sign-in' | 'link-current-account';

export type OidcStartResult =
  | {
      readonly kind: 'redirect';
      readonly request: OidcAuthorizationRequest;
    }
  | { readonly kind: 'failed'; readonly error: 'authentication-unavailable' };

export async function startGoogleOidc(input: {
  readonly configuration: unknown;
  readonly redirectUri: unknown;
  readonly intent: OidcStartIntent;
  readonly vaultContext?: VaultContext;
  readonly clock: OidcClockPort;
  readonly secrets: OidcSecretPort;
  readonly pkce: PkceChallengePort;
  readonly transactions: OidcTransactionStore;
}): Promise<OidcStartResult> {
  const configuration = oidcProviderConfigurationDecoder.decode(
    input.configuration,
  );
  const redirectUri = oidcRedirectUriDecoder.decode(input.redirectUri);
  const now = input.clock.nowEpochSeconds();
  if (
    !configuration.ok ||
    !redirectUri.ok ||
    typeof now !== 'number' ||
    !Number.isSafeInteger(now) ||
    now < 0 ||
    (input.intent === 'link-current-account' && !input.vaultContext)
  ) {
    return { kind: 'failed', error: 'authentication-unavailable' };
  }

  try {
    const [rawState, rawNonce, rawVerifier] = await Promise.all([
      input.secrets.createState(),
      input.secrets.createNonce(),
      input.secrets.createCodeVerifier(),
    ]);
    const state = oidcStateDecoder.decode(rawState);
    const nonce = oidcNonceDecoder.decode(rawNonce);
    const verifier = pkceCodeVerifierDecoder.decode(rawVerifier);
    if (!state.ok || !nonce.ok || !verifier.ok) {
      return { kind: 'failed', error: 'authentication-unavailable' };
    }
    const purpose = oidcPurposeFromStart(input.intent, input.vaultContext);
    if (!purpose) {
      return { kind: 'failed', error: 'authentication-unavailable' };
    }
    const transaction = createOidcTransaction({
      configuration: configuration.value,
      state: state.value,
      nonce: nonce.value,
      codeVerifier: verifier.value,
      redirectUri: redirectUri.value,
      purpose,
      nowEpochSeconds: now,
    });
    if (transaction.kind === 'rejected') {
      return { kind: 'failed', error: 'authentication-unavailable' };
    }
    const rawChallenge = await input.pkce.deriveS256(verifier.value);
    const challenge = pkceCodeChallengeDecoder.decode(rawChallenge);
    if (!challenge.ok) {
      return { kind: 'failed', error: 'authentication-unavailable' };
    }
    await input.transactions.insertPending(transaction.transaction);
    return {
      kind: 'redirect',
      request: createOidcAuthorizationRequest(
        configuration.value,
        transaction.transaction,
        challenge.value,
      ),
    };
  } catch {
    return { kind: 'failed', error: 'authentication-unavailable' };
  }
}

export type OidcCompletionResult =
  | {
      readonly kind: 'resolved';
      readonly resolution: Exclude<
        OidcIdentityResolution,
        { readonly kind: 'rejected' }
      >;
    }
  | { readonly kind: 'failed'; readonly error: 'authentication-failed' };

export async function completeGoogleOidc(input: {
  readonly callback: unknown;
  readonly configuration: unknown;
  readonly clock: OidcClockPort;
  readonly transactions: OidcTransactionStore;
  readonly provider: OidcVerifiedClaimsPort;
  readonly identities: OidcIdentityDirectory;
}): Promise<OidcCompletionResult> {
  const callback = oidcCallbackDecoder.decode(input.callback);
  const configuration = oidcProviderConfigurationDecoder.decode(
    input.configuration,
  );
  const now = input.clock.nowEpochSeconds();
  if (
    !callback.ok ||
    !configuration.ok ||
    typeof now !== 'number' ||
    !Number.isSafeInteger(now) ||
    now < 0
  ) {
    return authenticationFailed();
  }

  try {
    const rawTransaction = await input.transactions.consumeByState(
      callback.value.state,
    );
    const transaction = pendingOidcTransactionDecoder.decode(rawTransaction);
    if (!transaction.ok) return authenticationFailed();
    const transactionValidation = validateOidcTransaction(
      transaction.value,
      callback.value.state,
      configuration.value,
      now,
    );
    if (
      transactionValidation.kind === 'rejected' ||
      callback.value.kind === 'provider-error'
    ) {
      return authenticationFailed();
    }

    const rawClaims = await input.provider.exchangeCodeForVerifiedClaims({
      code: callback.value.code,
      clientId: configuration.value.clientId,
      redirectUri: transaction.value.redirectUri,
      codeVerifier: transaction.value.codeVerifier,
    });
    const claims = verifiedOidcClaimsDecoder.decode(rawClaims);
    if (!claims.ok) return authenticationFailed();
    const claimsValidation = validateOidcClaims(
      claims.value,
      transaction.value,
      configuration.value,
      now,
    );
    if (claimsValidation.kind === 'rejected') return authenticationFailed();

    const [rawIdentity, rawEmailAccountId] = await Promise.all([
      input.identities.findByIssuerSubject(claimsValidation.identityKey),
      input.identities.findAccountIdByVerifiedEmail(claimsValidation.email),
    ]);
    const existingIdentity = optionalIdentity(rawIdentity);
    const verifiedEmailAccountId = optionalAccountId(rawEmailAccountId);
    if (!existingIdentity.ok || !verifiedEmailAccountId.ok) {
      return authenticationFailed();
    }
    const resolution = decideOidcIdentityResolution({
      purpose: transaction.value.purpose,
      identityKey: claimsValidation.identityKey,
      email: claimsValidation.email,
      ...(existingIdentity.value === undefined
        ? {}
        : { existingIdentity: existingIdentity.value }),
      ...(verifiedEmailAccountId.value === undefined
        ? {}
        : { verifiedEmailAccountId: verifiedEmailAccountId.value }),
    });
    return resolution.kind === 'rejected'
      ? authenticationFailed()
      : { kind: 'resolved', resolution };
  } catch {
    return authenticationFailed();
  }
}

type OptionalDecoded<T> =
  | { readonly ok: true; readonly value: T | undefined }
  | { readonly ok: false };

function optionalIdentity(input: unknown): OptionalDecoded<OidcIdentityRecord> {
  if (input === undefined) return { ok: true, value: undefined };
  const decoded = oidcIdentityRecordDecoder.decode(input);
  return decoded.ok ? { ok: true, value: decoded.value } : { ok: false };
}

function optionalAccountId(input: unknown): OptionalDecoded<AccountId> {
  if (input === undefined) return { ok: true, value: undefined };
  const decoded = accountIdDecoder.decode(input);
  return decoded.ok ? { ok: true, value: decoded.value } : { ok: false };
}

function authenticationFailed(): OidcCompletionResult {
  return { kind: 'failed', error: 'authentication-failed' };
}

function oidcPurposeFromStart(
  intent: OidcStartIntent,
  context: VaultContext | undefined,
):
  | { readonly kind: 'sign-in' }
  | { readonly kind: 'link'; readonly accountId: AccountId }
  | undefined {
  if (intent === 'sign-in') return { kind: 'sign-in' };
  return context ? { kind: 'link', accountId: context.accountId } : undefined;
}

File: backend/internal/identity/oidc_boundary.go
package identity

import (
	"context"
	"net/url"
)

type OidcClock interface {
	NowEpochSeconds() int64
}

type OidcSecretPort interface {
	CreateState(context.Context) (string, error)
	CreateNonce(context.Context) (string, error)
	CreateCodeVerifier(context.Context) (string, error)
}

type PkceChallengePort interface {
	DeriveS256(PkceCodeVerifier) (string, error)
}

type OidcTransactionStore interface {
	InsertPending(context.Context, PendingOidcTransaction) error
	ConsumeByState(context.Context, OidcState) (*PendingOidcTransaction, error)
}

type OidcCodeExchangeInput struct {
	Code         OidcAuthorizationCode
	ClientID     OidcClientID
	RedirectURI  OidcRedirectURI
	CodeVerifier PkceCodeVerifier
}

type OidcVerifiedClaimsPort interface {
	ExchangeCodeForVerifiedClaims(context.Context, OidcCodeExchangeInput) (RawOidcClaims, error)
}

type OidcIdentityDirectory interface {
	FindByIssuerSubject(context.Context, OidcIdentityKey) (*OidcIdentityRecord, error)
	FindAccountIDByVerifiedEmail(context.Context, VerifiedEmailAddress) (*AccountID, error)
}

type OidcStartIntent string

const (
	OidcStartSignIn             OidcStartIntent = "sign-in"
	OidcStartLinkCurrentAccount OidcStartIntent = "link-current-account"
)

type OidcStartResult struct {
	Redirect bool
	Request  OidcAuthorizationRequest
	Error    string
}

func StartGoogleOidc(ctx context.Context, input struct {
	Configuration      OidcProviderConfiguration
	RedirectURI        string
	Intent             OidcStartIntent
	SignupTermsConsent *SignupTermsConsent
	VaultContext       *VaultContext
	Clock              OidcClock
	Secrets            OidcSecretPort
	Pkce               PkceChallengePort
	Transactions       OidcTransactionStore
}) OidcStartResult {
	failure := OidcStartResult{Error: "authentication-unavailable"}
	if input.Clock == nil || input.Secrets == nil || input.Pkce == nil || input.Transactions == nil ||
		!input.Configuration.Valid() {
		return failure
	}
	now := input.Clock.NowEpochSeconds()
	redirectURI, err := ParseOidcRedirectURI(input.RedirectURI)
	if err != nil || !validTimestamp(now) ||
		(input.Intent != OidcStartSignIn && input.Intent != OidcStartLinkCurrentAccount) ||
		(input.Intent == OidcStartLinkCurrentAccount &&
			(input.VaultContext == nil || !validOidcVaultContext(*input.VaultContext))) ||
		(input.Intent == OidcStartLinkCurrentAccount && input.SignupTermsConsent != nil) ||
		(input.SignupTermsConsent != nil && !input.SignupTermsConsent.Valid()) {
		return failure
	}
	rawState, err := input.Secrets.CreateState(ctx)
	if err != nil {
		return failure
	}
	rawNonce, err := input.Secrets.CreateNonce(ctx)
	if err != nil {
		return failure
	}
	rawVerifier, err := input.Secrets.CreateCodeVerifier(ctx)
	if err != nil {
		return failure
	}
	state, err := ParseOidcState(rawState)
	if err != nil {
		return failure
	}
	nonce, err := ParseOidcNonce(rawNonce)
	if err != nil {
		return failure
	}
	verifier, err := ParsePkceCodeVerifier(rawVerifier)
	if err != nil {
		return failure
	}
	purpose := OidcPurpose{Kind: OidcPurposeSignIn}
	if input.Intent == OidcStartLinkCurrentAccount {
		purpose = OidcPurpose{Kind: OidcPurposeLink, AccountID: input.VaultContext.AccountID}
	}
	decision := CreateOidcTransaction(PendingOidcTransaction{
		State: state, Nonce: nonce, CodeVerifier: verifier, RedirectURI: redirectURI,
		Purpose: purpose, SignupTermsConsent: input.SignupTermsConsent,
		CreatedAtEpochSeconds: now,
	}, input.Configuration)
	if !decision.Created {
		return failure
	}
	rawChallenge, err := input.Pkce.DeriveS256(verifier)
	if err != nil {
		return failure
	}
	challenge, err := ParsePkceCodeChallenge(rawChallenge)
	if err != nil {
		return failure
	}
	if err := input.Transactions.InsertPending(ctx, decision.Transaction); err != nil {
		return failure
	}
	return OidcStartResult{
		Redirect: true,
		Request:  CreateOidcAuthorizationRequest(input.Configuration, decision.Transaction, challenge),
	}
}

func validOidcVaultContext(value VaultContext) bool {
	return validAccountID(value.AccountID) && validVaultID(value.VaultID) &&
		validSessionID(value.SessionID) && validEpoch(value.SessionEpoch)
}

type OidcCallbackKind string

const (
	OidcCallbackAuthorizationCode OidcCallbackKind = "authorization-code"
	OidcCallbackProviderError     OidcCallbackKind = "provider-error"
)

type OidcCallback struct {
	Kind          OidcCallbackKind
	State         OidcState
	Code          OidcAuthorizationCode
	ProviderError string
}

type OidcCallbackInput struct {
	State            string
	Code             *string
	ProviderError    *string
	ErrorDescription *string
}

func DecodeOidcCallback(input OidcCallbackInput) (OidcCallback, error) {
	state, err := ParseOidcState(input.State)
	if err != nil || (input.Code == nil) == (input.ProviderError == nil) ||
		(input.ProviderError == nil && input.ErrorDescription != nil) {
		return OidcCallback{}, ErrInvalidOidcValue
	}
	if input.Code != nil {
		code, parseErr := ParseOidcAuthorizationCode(*input.Code)
		if parseErr != nil {
			return OidcCallback{}, parseErr
		}
		return OidcCallback{Kind: OidcCallbackAuthorizationCode, State: state, Code: code}, nil
	}
	if len(*input.ProviderError) < 1 || len(*input.ProviderError) > 256 ||
		!visibleASCIIPattern.MatchString(*input.ProviderError) ||
		(input.ErrorDescription != nil && len(*input.ErrorDescription) > 1_024) {
		return OidcCallback{}, ErrInvalidOidcValue
	}
	return OidcCallback{
		Kind: OidcCallbackProviderError, State: state, ProviderError: *input.ProviderError,
	}, nil
}

func OidcCallbackInputFromURL(raw string) (OidcCallbackInput, error) {
	if len(raw) > 8_192 {
		return OidcCallbackInput{}, ErrInvalidOidcValue
	}
	parsed, err := url.Parse(raw)
	if err != nil {
		return OidcCallbackInput{}, ErrInvalidOidcValue
	}
	values, err := url.ParseQuery(parsed.RawQuery)
	if err != nil {
		return OidcCallbackInput{}, ErrInvalidOidcValue
	}
	for _, name := range []string{"state", "code", "error", "error_description"} {
		if len(values[name]) > 1 {
			return OidcCallbackInput{}, ErrInvalidOidcValue
		}
	}
	input := OidcCallbackInput{State: values.Get("state")}
	if values.Has("code") {
		value := values.Get("code")
		input.Code = &value
	}
	if values.Has("error") {
		value := values.Get("error")
		input.ProviderError = &value
	}
	if values.Has("error_description") {
		value := values.Get("error_description")
		input.ErrorDescription = &value
	}
	return input, nil
}

func SerializeOidcAuthorizationRequest(request OidcAuthorizationRequest) (string, error) {
	endpoint, err := url.Parse(string(request.AuthorizationEndpoint))
	if err != nil {
		return "", ErrInvalidOidcValue
	}
	query := endpoint.Query()
	query.Set("response_type", request.ResponseType)
	query.Set("client_id", string(request.ClientID))
	query.Set("redirect_uri", string(request.RedirectURI))
	query.Set("scope", request.Scope)
	query.Set("state", string(request.State))
	query.Set("nonce", string(request.Nonce))
	query.Set("code_challenge", string(request.CodeChallenge))
	query.Set("code_challenge_method", request.CodeChallengeMethod)
	endpoint.RawQuery = query.Encode()
	return endpoint.String(), nil
}

type OidcCompletionKind string

const (
	OidcCompletionResolved OidcCompletionKind = "resolved"
	OidcCompletionAdmitted OidcCompletionKind = "admitted"
	OidcCompletionFailed   OidcCompletionKind = "failed"
)

type OidcCompletionResult struct {
	Kind       OidcCompletionKind
	Error      string
	Resolution OidcIdentityResolution
	Admission  SignupAdmissionReceipt
}

func CompleteGoogleOidc(ctx context.Context, input struct {
	Callback      OidcCallbackInput
	Configuration OidcProviderConfiguration
	Clock         OidcClock
	Transactions  OidcTransactionStore
	Provider      OidcVerifiedClaimsPort
	Identities    OidcIdentityDirectory
	Signup        SignupAdmissionPort
}) OidcCompletionResult {
	failure := OidcCompletionResult{Kind: OidcCompletionFailed, Error: "authentication-failed"}
	if input.Clock == nil || input.Transactions == nil || input.Provider == nil || input.Identities == nil ||
		!input.Configuration.Valid() {
		return failure
	}
	callback, err := DecodeOidcCallback(input.Callback)
	if err != nil {
		return failure
	}
	now := input.Clock.NowEpochSeconds()
	if !validTimestamp(now) {
		return failure
	}
	transaction, err := input.Transactions.ConsumeByState(ctx, callback.State)
	if err != nil || transaction == nil || !ValidPendingOidcTransaction(*transaction) {
		return failure
	}
	if validation := ValidateOidcTransaction(*transaction, callback.State, input.Configuration, now); !validation.Valid ||
		callback.Kind == OidcCallbackProviderError {
		return failure
	}
	rawClaims, err := input.Provider.ExchangeCodeForVerifiedClaims(ctx, OidcCodeExchangeInput{
		Code: callback.Code, ClientID: input.Configuration.ClientID,
		RedirectURI: transaction.RedirectURI, CodeVerifier: transaction.CodeVerifier,
	})
	if err != nil {
		return failure
	}
	claims, err := DecodeVerifiedOidcClaims(rawClaims)
	if err != nil {
		return failure
	}
	claimsValidation := ValidateOidcClaims(claims, *transaction, input.Configuration, now)
	if !claimsValidation.Valid {
		return failure
	}
	existingIdentity, err := input.Identities.FindByIssuerSubject(ctx, claimsValidation.IdentityKey)
	if err != nil || (existingIdentity != nil && !existingIdentity.Valid()) {
		return failure
	}
	verifiedEmailAccountID, err := input.Identities.FindAccountIDByVerifiedEmail(ctx, claimsValidation.Email.Verified())
	if err != nil || (verifiedEmailAccountID != nil && !validAccountID(*verifiedEmailAccountID)) {
		return failure
	}
	resolution := DecideOidcIdentityResolution(struct {
		Purpose                OidcPurpose
		IdentityKey            OidcIdentityKey
		Email                  OidcEmailAddress
		ExistingIdentity       *OidcIdentityRecord
		VerifiedEmailAccountID *AccountID
	}{
		Purpose: transaction.Purpose, IdentityKey: claimsValidation.IdentityKey,
		Email: claimsValidation.Email, ExistingIdentity: existingIdentity,
		VerifiedEmailAccountID: verifiedEmailAccountID,
	})
	if resolution.Kind == OidcIdentityRejected {
		return failure
	}
	if resolution.Kind == OidcProvisionAccount {
		if transaction.SignupTermsConsent == nil || input.Signup == nil {
			return failure
		}
		verifiedIdentity := VerifiedSignupIdentity{
			Kind:   SignupIdentityGoogle,
			Issuer: resolution.IdentityKey.Issuer, Subject: resolution.IdentityKey.Subject,
			Email: resolution.Email.Verified(),
		}
		admission, admitErr := input.Signup.Admit(ctx, verifiedIdentity, *transaction.SignupTermsConsent)
		if admitErr != nil || !admission.Admitted ||
			!signupAdmissionMatchesRequest(admission.Receipt, verifiedIdentity, *transaction.SignupTermsConsent) {
			return failure
		}
		return OidcCompletionResult{Kind: OidcCompletionAdmitted, Admission: admission.Receipt}
	}
	return OidcCompletionResult{Kind: OidcCompletionResolved, Resolution: resolution}
}
